sway-core/src/semantic_analysis/cei_pattern_analysis.rs

```diff
@@ -0,0 +1,258 @@
+// CEI stands for "Checks, Effects, Interactions".  We check that no storage writes
+// (effects) occur after calling external contracts (interaction) and issue warnings
+// if it's the case.
+// See this [blog post](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html)
+// for more detail.
+
+use crate::{
+    declaration_engine::DeclarationId,
+    language::ty::{self, TyFunctionDeclaration},
+};
+use std::collections::HashSet;
+use sway_error::warning::{CompileWarning, Warning};
+use sway_types::Span;
+
+#[derive(PartialEq, Eq, Hash, Clone)]
+enum Effect {
+    Interaction,  // interaction with external contracts
+    StorageWrite, // storage modification
+}
+
+enum CEIAnalysisState {
+    LookingForInteraction,
+    LookingForStorageWrite,
+}
+
+// TODO: process libraries, scripts and predicates
+pub(crate) fn program(prog: &ty::TyProgram) -> Vec<CompileWarning> {
+    match &prog.kind {
+        ty::TyProgramKind::Library { .. } => vec![],
+        ty::TyProgramKind::Script { .. } => vec![],
+        ty::TyProgramKind::Predicate { .. } => vec![],
```

Comment:
The good news is that you don't have to worry about libraries, scripts, or predicates because they can't access storage! Libraries may access storage in the future, but we'll cross that bridge when we understand how that looks like.